File: scripts/cobertura_clm.py
"""Calculo de cobertura CLM (educacion, admin) — compartido por informe y tests."""
from __future__ import annotations

import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]

CUERPO_SLUG = {
    "0597": "maestros",
    "0590": "secundaria",
    "0591": "tecnicos-fp",
    "0592": "eoii",
    "0593": "catedraticos-musica",
    "0594": "profesores-musica",
    "0595": "artes-plasticas",
    "0596": "maestros-taller",
    "0598": "fp-singulares",
}

METADATA = frozenset({"manifest.json", "categorias.json", "afinidad.json"})


def slug_archivo(texto: str) -> str:
    import unicodedata

    s = unicodedata.normalize("NFD", texto)
    s = "".join(c for c in s if unicodedata.category(c) != "Mn")
    s = s.lower().replace("/", "-")
    s = re.sub(r"[^a-z0-9]+", "-", s)
    return s.strip("-")


def contar_listados_manifest(manifest: dict | None) -> int:
    archivos = (manifest or {}).get("archivos") or []
    return sum(
        1
        for a in archivos
        if a.endswith(".json")
        and not a.endswith(".busqueda.json")
        and a.split("/")[-1] not in METADATA
    )


def _rel(cuerpo_codigo: str, esp: str) -> str | None:
    g = CUERPO_SLUG.get(cuerpo_codigo)
    if not g:
        return None
    m = re.match(r"^(\d{3})\s+(.+)$", esp)
    if not m:
        return None
    return f"{g}/{slug_archivo(f'{m.group(1)}-{m.group(2)}')}.json"


def educacion_huecos(
    categorias_path: Path | None = None,
    man_d_path: Path | None = None,
    man_b_path: Path | None = None,
) -> dict:
    categorias_path = categorias_path or ROOT / "data/educacion/categorias.json"
    man_d_path = man_d_path or ROOT / "data/educacion/manifest.json"
    man_b_path = man_b_path or ROOT / "data/educacion-bolsa/manifest.json"

    cat = json.loads(categorias_path.read_text(encoding="utf-8"))
    man_d = set()
    man_b = set()
    if man_d_path.is_file():
        man_d = set(json.loads(man_d_path.read_text(encoding="utf-8")).get("archivos") or [])
    if man_b_path.is_file():
        man_b = set(json.loads(man_b_path.read_text(encoding="utf-8")).get("archivos") or [])

    sin_d, sin_b = [], []
    catalogo = 0
    for cuerpo in cat.get("cuerpos", []):
        cod = cuerpo["codigo"]
        for esp in cuerpo.get("especialidades", []):
            catalogo += 1
            rel = _rel(cod, esp)
            if not rel:
                continue
            item = {"cuerpo": cuerpo["nombre"], "especialidad": esp, "rel": rel}
            if rel not in man_d:
                sin_d.append(item)
            if rel not in man_b:
                sin_b.append(item)

    return {
        "catalogo": catalogo,
        "disponibles": contar_listados_manifest({"archivos": list(man_d)}),
        "bolsa": contar_listados_manifest({"archivos": list(man_b)}),
        "faltantes_disponibles": sin_d,
        "faltantes_bolsa": sin_b,
    }


def admin_sin_pdf(categorias_path: Path | None = None) -> list[dict]:
    categorias_path = categorias_path or ROOT / "data/admin-clm/categorias.json"
    if not categorias_path.is_file():
        return []
    data = json.loads(categorias_path.read_text(encoding="utf-8"))
    return [
        {
            "categoria": e.get("categoria"),
            "colectivo": e.get("colectivo"),
            "nota": e.get("nota_portal"),
            "url": e.get("url_pagina"),
        }
        for e in data
        if e.get("sin_pdf_portal")
    ]

File: src/cobertura-clm.js
/** Cobertura de datos CLM y helpers de frescura para la UI. */
import { CUERPO_SLUG } from "./educacion.js";

const METADATA_JSON = new Set(["manifest.json", "categorias.json", "afinidad.json"]);

export function contarListadosManifest(manifest) {
  return (manifest?.archivos || []).filter(
    (a) =>
      a.endsWith(".json") &&
      !a.endsWith(".busqueda.json") &&
      !METADATA_JSON.has(a.split("/").pop()),
  ).length;
}

function relEspecialidad(cuerpoCodigo, esp, slugArchivo) {
  const grupoId = CUERPO_SLUG[cuerpoCodigo];
  if (!grupoId) return null;
  const m = esp.match(/^(\d{3})\s+(.+)$/);
  if (!m) return null;
  return `${grupoId}/${slugArchivo(`${m[1]}-${m[2]}`)}.json`;
}

/** Compara catalogo educacion con manifests disponibles y bolsa. */
export function calcularCoberturaEducacion(categoriasDoc, manifestDisponibles, manifestBolsa, slugArchivo) {
  const archivosD = new Set(manifestDisponibles?.archivos || []);
  const archivosB = new Set(manifestBolsa?.archivos || []);
  let catalogo = 0;
  const faltantesDisponibles = [];
  const faltantesBolsa = [];

  for (const cuerpo of categoriasDoc?.cuerpos || []) {
    for (const esp of cuerpo.especialidades || []) {
      catalogo += 1;
      const rel = relEspecialidad(cuerpo.codigo, esp, slugArchivo);
      if (!rel) continue;
      const base = {
        cuerpo: cuerpo.nombre,
        codigoCuerpo: cuerpo.codigo,
        especialidad: esp,
        rel,
      };
      if (!archivosD.has(rel)) faltantesDisponibles.push(base);
      if (!archivosB.has(rel)) faltantesBolsa.push(base);
    }
  }

  return {
    catalogo,
    disponibles: contarListadosManifest(manifestDisponibles),
    bolsa: contarListadosManifest(manifestBolsa),
    faltantesDisponibles,
    faltantesBolsa,
    urlPortal:
      categoriasDoc?.fuente?.pagina_bolsas ||
      "https://educacion.castillalamancha.es/profesorado/bolsas-de-trabajo",
  };
}

export function adminBolsasSinPdf(categoriasList) {
  return (categoriasList || [])
    .filter((e) => e.sin_pdf_portal)
    .map((e) => ({
      categoria: e.categoria,
      colectivo: e.colectivo,
      nota: e.nota_portal || "Sin PDF de listado en el portal.",
      url: e.url_pagina || null,
    }));
}

export function frescuraDesdeManifests(manifests) {
  return {
    sanidad: manifests.sanidad?.generado || null,
    educacionDisponibles: manifests.educacion?.generado || null,
    educacionBolsa: manifests.educacionBolsa?.generado || null,
    admin: manifests.admin?.generado || null,
  };
}

export function formatearFrescura(iso) {
  if (!iso) return null;
  try {
    const d = new Date(iso);
    if (Number.isNaN(d.getTime())) return iso;
    return d.toLocaleString("es-ES", {
      day: "numeric",
      month: "short",
      year: "numeric",
      hour: "2-digit",
      minute: "2-digit",
    });
  } catch {
    return iso;
  }
}

/** Texto corto de frescura para el sector activo en buscar. */
export function etiquetaFrescuraSector(sectorId, frescura, modoListadoEducacion) {
  if (!frescura) return null;
  if (sectorId === "sanidad") return formatearFrescura(frescura.sanidad);
  if (sectorId === "administracion") return formatearFrescura(frescura.admin);
  if (sectorId === "educacion") {
    if (modoListadoEducacion === "disponibles") return formatearFrescura(frescura.educacionDisponibles);
    return formatearFrescura(frescura.educacionBolsa);
  }
  return null;
}
